### PR TITLE
[SYCL][ESIMD] Implement prefetch API that accepting compile time properties

### PR DESCRIPTION
**Pull request by fineg74**

### CODE REVIEW

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 1]
```diff
@@ -8004,6 +8069,338 @@ scatter(AccessorTy acc, simd<uint32_t, N> offsets, simd<T, N> vals,
...
+}
+
+/// template <typename T, int VS = 1, typename OffsetT,
+///           typename PropertyListT = empty_properties_t>
+/// void prefetch(const T *p, OffsetT byte_offset, simd_mask<1> mask,
+///                   PropertyListT props = {});                   // (usm-pf-5)
+/// Supported platforms: DG2, PVC only.
+/// Prefetches elements of the type 'T' from continuous memory location
+/// addressed by the base pointer \p p, and offset \p byte_offset and the length
+/// \p VS into the cache.
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:8245]:
```suggestion
/// \p VS elements into the cache.
```

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 2]
```diff
@@ -8004,6 +8069,338 @@ scatter(AccessorTy acc, simd<uint32_t, N> offsets, simd<T, N> vals,
...
+/// @tparam T Element type.
+/// @tparam VS Vector size. It can also be read as the number of reads per each
+/// address. The parameter 'N' must be divisible by 'VS'.
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:8248]: This does not seem accurate for transposed prefetch. Also, this function does not have the parameter parameter 'N'.
Maybe something like this?
```suggestion
/// @tparam VS Vector size. It specifies the number of consequent elements to prefetch.
```

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 3]
```diff
@@ -8004,6 +8069,338 @@ scatter(AccessorTy acc, simd<uint32_t, N> offsets, simd<T, N> vals,
...
+///                   PropertyListT props = {});                   // (usm-pf-7)
+/// void prefetch(const T *p, PropertyListT props = {});           // (usm-pf-8)
+///
+/// The next 2 functions are variations of the first 2 above (usm-pf-1,2)
+/// and were added only to support simd_view instead of simd for byte_offsets
+/// operand.
+/// template <typename T, int N, int VS = 1, typename OffsetT,
+///           typename PropertyListT = empty_properties_t>
+/// void prefetch(const T *p,
+///             simd_view<OffsetObjT, OffsetRegionT> byte_offsets,
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:8110]: The lines here use `simd_view<OffsetObjT, OffsetRegionT>`, while the actual definition uses template `typename OffsetSimdViewT`.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 4]
```diff
@@ -8004,6 +8069,338 @@ scatter(AccessorTy acc, simd<uint32_t, N> offsets, simd<T, N> vals,
...
+/// omitted.
+/// @param props The optional compile-time properties. Only  cache hint
+/// properties are used.
+template <typename T, int VS = 1, typename OffsetT,
+          typename PropertyListT =
+              ext::oneapi::experimental::detail::empty_properties_t>
+__ESIMD_API std::enable_if_t<
+    std::is_integral_v<OffsetT> &&
+    ext::oneapi::experimental::is_property_list_v<PropertyListT>>
+prefetch(const T *p, OffsetT byte_offset, simd_mask<1> mask,
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:8261]: Does this function co-exists with (usm-pf-3) when there are calls like `prefetch<float, 16>(ptr, 0, 1)`?

File: sycl/test/esimd/memory_properties.cpp [hunk 5]
```diff
@@ -1453,3 +1455,113 @@ test_slm_gather_scatter(int byte_offset32) {
...
+
+  prefetch<float, 32, 2>(ptrf, loffset_n16, mask_n16);
+  prefetch<float, 32, 2>(ptrf, loffset_n16_view, mask_n16);
+
+  // CHECK-COUNT-4: call void @llvm.genx.lsc.prefetch.stateless.v16i1.v16i64(<16 x i1> {{[^)]+}}, i8 0, i8 2, i8 1, i16 1, i32 0, i8 3, i8 2, i8 1, i8 0, <16 x i64> {{[^)]+}}, i32 0)
+  prefetch<float, 32, 2>(ptrf, ioffset_n16, mask_n16, props_cache_load);
+  prefetch<float, 32, 2>(ptrf, ioffset_n16_view, mask_n16, props_cache_load);
+
+  prefetch<float, 32, 2>(ptrf, loffset_n16, mask_n16, props_cache_load);
+  prefetch<float, 32, 2>(ptrf, loffset_n16_view, mask_n16, props_cache_load);
```
Comment (v-klochkov) [inline sycl/test/esimd/memory_properties.cpp:1566]: Please add some more test cases like `prefetch<float, 16>(ptr, 0)` and `prefetch<float, 16>(ptr, 0, 1)` trying to create a situation of ambiguous call error.

**Review by v-klochkov — COMMENTED** [review]
I am still reviewing the patch. Sending here the comments accumulated to this moment.

File: sycl/test-e2e/ESIMD/unified_memory_api/Inputs/prefetch.hpp [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+       if constexpr (VS > 1) { // VS > 1 requires specifying <T, N, VS>
+         if constexpr (UseMask) {
+           if constexpr (UseProperties) {
+             if constexpr (sizeof(T) >= 4) {
+               if (GlobalID % 2 == 0) // ByteOffset - simd
+                 prefetch<T, N, VS>(In, ByteOffsets, Pred, Props);
+               else if (GlobalID % 4 == 0)
+                 __ESIMD_NS::prefetch<T, VS>(In, Pred_1, Props);
+               else // ByteOffset - simd_view
+                 prefetch<T, N, VS>(In, ByteOffsetsView, Pred, Props);
```
Comment (v-klochkov) [inline sycl/test-e2e/ESIMD/unified_memory_api/Inputs/prefetch.hpp:99]: these 3 calls test (usm-pf,1,9,5). Can you please add the test case for (usm-pf-7), i.e. prefetch<T, VS>(In, ScalarByteOffset, Pred_1, Props)

**Review by v-klochkov — COMMENTED** [review]
I finished the code-review - it looks good. I have just few more comments.
@sarnex - please take a look at this patch too.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 7]
```diff
@@ -8004,6 +8065,335 @@ scatter(AccessorTy acc, simd<uint32_t, N> offsets, simd<T, N> vals,
...
+/// and were added only to support simd_view instead of simd for byte_offsets
+/// operand.
+/// template <typename T, int N, int VS = 1, typename OffsetSimdViewT,
+///           typename PropertyListT = empty_properties_t>
+/// void prefetch(const T *p, OffsetSimdViewT byte_offsets,
+///             simd_mask<N / VS> mask, PropertyListT props = {}); // (usm-pf-5)
+/// void prefetch(const T *p, OffsetSimdViewT byte_offsets,
+///             PropertyListT props = {});                        // (usm-pf-6)
+///
+/// The next functions perform transposed 1-channel prefetch
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:8096]:
```suggestion
/// The next functions perform transposed 1-channel prefetch.
```

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 8]
```diff
@@ -8004,6 +8065,335 @@ scatter(AccessorTy acc, simd<uint32_t, N> offsets, simd<T, N> vals,
...
+/// addressed by the base pointer \p p, and offset \p byte_offset and the length
+/// \p VS elements into the cache.
+/// @tparam T Element type.
+/// @tparam VS Vector size. It specifies the number of consequent elements to
+/// prefetch.
+/// @param p The base address.
+/// @param byte_offset offset from the base address.
+/// @param mask The access mask. If it is set to 0, then the prefetch is
+/// omitted.
+/// @param props The optional compile-time properties. Only  cache hint
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:8304]:
```suggestion
/// @param props The optional compile-time properties. Only cache hint
```

nit: looks like there's an extra space after 'Only' here and in some others

File: sycl/test-e2e/ESIMD/unified_memory_api/prefetch_usm_dg2_pvc.cpp [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// REQUIRES: gpu-intel-dg2 || gpu-intel-pvc
+// RUN: %{build} -fsycl-device-code-split=per_kernel -o %t.out
+// RUN: %{run} %t.out
+
+// The test verifies esimd::prefetch() functions accepting USM pointer
+// and optional compile-time esimd::properties.
+// The prefetch() calls in this test  require DG2 or PVC to run.
```
Comment (sarnex) [inline sycl/test-e2e/ESIMD/unified_memory_api/prefetch_usm_dg2_pvc.cpp:14]:
```suggestion
// The prefetch() calls in this test require DG2 or PVC to run.
```
nit: extra space

**Review by sarnex — APPROVED** [review]
LGTM, just super minor stuff

**Review by v-klochkov — APPROVED** [review]
Looks good.
We need to think about support of 64x64-bytes transposed prefetch, similarly to how block_load<float, 128> is implemented via usage of i64x64 on PVC. 

Doing so, would require accepting not only cache hints, but the alignment property too. Not sure if transposed prefetch requires the alignment to element type, need to check with GPU requirements.

**Review by sarnex — APPROVED** [review]

**Pull request merged**